orttraining/orttraining/training_ops/cuda/communication/send.cc

```diff
@@ -3,69 +3,130 @@
+        .OutputMemoryType<OrtMemTypeCPUOutput>(0) /* CPU variable */
+        .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>())
+        .TypeConstraint("TInt64", DataTypeImpl::GetTensorType<int64_t>())
+        .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
+    Send);
 
 void CUDART_CB HostSend(void* args) {
   CommInfo_t* info = reinterpret_cast<CommInfo_t*>(args);
-  int mpi_code  = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
+  int mpi_code = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
   ORT_ENFORCE(mpi_code == MPI_SUCCESS, "MPI Send fails.");
 }
 
-template <typename T>
-Status Send<T>::ComputeInternal(OpKernelContext* ctx) const {
+Status Send::ComputeInternal(OpKernelContext* ctx) const {
   // Check if control signal is true.
   const Tensor* input_signal_tensor = ctx->Input<Tensor>(0);
   const bool* input_signal = input_signal_tensor->template Data<bool>();
   ORT_ENFORCE(*input_signal, "Input control signal of Send must be true before executing the node.");
 
-  // Start the communication.
-  const Tensor* x_tensor = ctx->Input<Tensor>(1);
-  size_t tensor_shape_size = static_cast<size_t>(x_tensor->Shape().NumDimensions());;
-  size_t tensor_size = static_cast<size_t>(x_tensor->Shape().Size());
-
+  // Extract remote rank
+  const Tensor* remote_rank_tensor = ctx->Input<Tensor>(1);
+  const int64_t* remote_rank = remote_rank_tensor->template Data<int64_t>();
+  const int dst = static_cast<int>(*remote_rank);
+
+  // Create buffers
+  const int tensor_num = static_cast<int>(element_types_.size());
+  // TODO move the following variables to member variables for extending life-time
+  // if we want to make the entire call async
+  std::vector<size_t> prefix_tensor_shape_sizes;
+  std::vector<int64_t> aggregated_tensor_shapes;
+  size_t aggregated_aligned_tensor_bytes = 0;
+  // tensor_offsets_in_bytes[i] is the starting byte of the i-th tensor in the send tensor buffer
+  std::vector<size_t> tensor_offsets_in_bytes;
+  // tensor_sizes_in_bytes[i] = (# of elements in the i-th tensor) * sizeof(the i-th tensor's element type)
+  std::vector<size_t> tensor_sizes_in_bytes;
+
+  // Compute tensor shapes and sizes
+  size_t prefix_tensor_shape_size_sum = 0;
+  for (int i = 0; i < tensor_num; ++i) {
+    const Tensor* x_tensor = ctx->Input<Tensor>(i + 2);
+    prefix_tensor_shape_size_sum += x_tensor->Shape().NumDimensions();
+    prefix_tensor_shape_sizes.push_back(prefix_tensor_shape_size_sum);
+    aggregated_tensor_shapes.insert(aggregated_tensor_shapes.end(),
+                                    x_tensor->Shape().GetDims().begin(),
+                                    x_tensor->Shape().GetDims().end());
+
+    // Find the next aligned offset in the tensor buffer to meet alignment requirement
+    aggregated_aligned_tensor_bytes = GetAggregatedAlignedAddress(aggregated_aligned_tensor_bytes);
```

Comment:
>GetAggregatedAlignedAddress [](start = 38, length = 27)

Why we need to do a per tensor alignment? For memcpy? 

I wonder if this is really needed...

Reply:
The aggregated tensor has to be copied to CPU, but only one allocation is called. The alignment needed to be guarded by each tensor for peak perf.